gno.land/pkg/gnoweb/url.go

```diff
@@ -4,145 +4,192 @@ import (
 	"errors"
 	"fmt"
 	"net/url"
+	"path/filepath"
 	"regexp"
 	"strings"
 )
 
-type PathKind byte
+var ErrURLInvalidPath = errors.New("invalid path")
 
-const (
-	KindInvalid PathKind = 0
-	KindRealm   PathKind = 'r'
-	KindPure    PathKind = 'p'
-)
+// rePkgOrRealmPath matches and validates a flexible path.
+var rePkgOrRealmPath = regexp.MustCompile(`^/[a-z][a-z0-9_/]*$`)
 
 // GnoURL decomposes the parts of an URL to query a realm.
 type GnoURL struct {
 	// Example full path:
-	// gno.land/r/demo/users:jae$help&a=b?c=d
+	// gno.land/r/demo/users/render.gno:jae$help&a=b?c=d
 
 	Domain   string     // gno.land
 	Path     string     // /r/demo/users
 	Args     string     // jae
 	WebQuery url.Values // help&a=b
 	Query    url.Values // c=d
+	File     string     // render.gno
 }
 
-func (url GnoURL) EncodeArgs() string {
-	var urlstr strings.Builder
-	if url.Args != "" {
-		urlstr.WriteString(url.Args)
-	}
+// EncodeFlag is used to specify which URL components to encode.
+type EncodeFlag int
 
-	if len(url.Query) > 0 {
-		urlstr.WriteString("?" + url.Query.Encode())
-	}
+const (
+	EncodePath     EncodeFlag = 1 << iota // Encode the path component
+	EncodeArgs                            // Encode the arguments component
+	EncodeWebQuery                        // Encode the web query component
+	EncodeQuery                           // Encode the query component
+	EncodeNoEscape                        // Disable escaping of arguments
+)
 
-	return urlstr.String()
+// Has checks if the EncodeFlag contains all the specified flags.
+func (f EncodeFlag) Has(flags EncodeFlag) bool {
+	return f&flags != 0
 }
 
-func (url GnoURL) EncodePath() string {
+// Encode encodes the URL components based on the provided flags.
+func (gnoURL GnoURL) Encode(encodeFlags EncodeFlag) string {
```

Comment:
What do you think about adding a small section on the `gnoweb` docs for these URL flags?

Reply:
b3c9463eaa5926cc0cb1eb3da0524d88d91c4614